File: ui/src/types/events.ts
/**
 * Frontend event type definitions matching backend event dataclasses.
 *
 * Maps 1:1 to the Python event types in `src/policy_factory/events.py`.
 * The discriminated union `PolicyEvent` enables type-safe event handling
 * in the central dispatcher.
 *
 * Events arrive over WebSocket with:
 * - `db_id` — database-generated integer (for deduplication and replay cursor)
 * - `id` — UUID string (original event ID)
 * - `event_type` — string literal discriminator
 * - `timestamp` — ISO 8601 string
 * - Type-specific fields
 *
 * The replay endpoint returns events with the integer `id` field (db_id).
 */

// ── Event type literals ────────────────────────────────────────────

export type EventType =
  // Cascade lifecycle (7)
  | "cascade_started"
  | "cascade_completed"
  | "cascade_failed"
  | "cascade_paused"
  | "cascade_resumed"
  | "cascade_cancelled"
  | "cascade_queued"
  // Layer processing (6)
  | "layer_generation_started"
  | "layer_generation_completed"
  | "critic_started"
  | "critic_completed"
  | "synthesis_started"
  | "synthesis_completed"
  // Agent streaming (1)
  | "agent_text_chunk"
  // Heartbeat (3)
  | "heartbeat_started"
  | "heartbeat_tier_completed"
  | "heartbeat_completed"
  // Seed progress (3)
  | "seed_started"
  | "seed_progress"
  | "seed_completed"
  // Ideas (5)
  | "idea_submitted"
  | "idea_evaluation_started"
  | "idea_evaluation_completed"
  | "idea_generation_started"
  | "idea_generation_completed"
  // System (4)
  | "user_login"
  | "user_created"
  | "cascade_lock_acquired"
  | "cascade_lock_released"
  // Conversation (6)
  | "conversation_started"
  | "conversation_text_chunk"
  | "conversation_file_edit"
  | "conversation_turn_complete"
  | "conversation_cascade_pending"
  | "conversation_turn_error";

export type EventCategory = "cascade" | "conversation" | "heartbeat" | "idea" | "seed" | "system";

// ── Base event interface ───────────────────────────────────────────

export interface BaseEvent {
  /** Database-generated integer ID (for deduplication and replay cursor).
   * Present on WS events as `db_id`, on replay events as `id`. */
  db_id: number;
  /** UUID string — the original event ID */
  id: string;
  event_type: EventType;
  timestamp: string; // ISO 8601
}

// ── Cascade lifecycle events ───────────────────────────────────────

export interface CascadeStartedEvent extends BaseEvent {
  event_type: "cascade_started";
  cascade_id: string;
  trigger_source: string;
  starting_layer: string;
}

export interface CascadeCompletedEvent extends BaseEvent {
  event_type: "cascade_completed";
  cascade_id: string;
  success: boolean;
}

export interface CascadeFailedEvent extends BaseEvent {
  event_type: "cascade_failed";
  cascade_id: string;
  error: string;
  failed_layer: string;
  failed_step: string;
}

export interface CascadePausedEvent extends BaseEvent {
  event_type: "cascade_paused";
  cascade_id: string;
  error: string;
  paused_layer: string;
  paused_step: string;
}

export interface CascadeResumedEvent extends BaseEvent {
  event_type: "cascade_resumed";
  cascade_id: string;
}

export interface CascadeCancelledEvent extends BaseEvent {
  event_type: "cascade_cancelled";
  cascade_id: string;
}

export interface CascadeQueuedEvent extends BaseEvent {
  event_type: "cascade_queued";
  cascade_id: string;
  queue_position: number;
}

// ── Layer processing events ────────────────────────────────────────

export interface LayerGenerationStartedEvent extends BaseEvent {
  event_type: "layer_generation_started";
  cascade_id: string;
  layer_slug: string;
}

export interface LayerGenerationCompletedEvent extends BaseEvent {
  event_type: "layer_generation_completed";
  cascade_id: string;
  layer_slug: string;
}

export interface CriticStartedEvent extends BaseEvent {
  event_type: "critic_started";
  cascade_id: string;
  layer_slug: string;
  critic_archetype: string;
}

export interface CriticCompletedEvent extends BaseEvent {
  event_type: "critic_completed";
  cascade_id: string;
  layer_slug: string;
  critic_archetype: string;
}

export interface SynthesisStartedEvent extends BaseEvent {
  event_type: "synthesis_started";
  cascade_id: string;
  layer_slug: string;
}

export interface SynthesisCompletedEvent extends BaseEvent {
  event_type: "synthesis_completed";
  cascade_id: string;
  layer_slug: string;
}

// ── Agent streaming events ─────────────────────────────────────────

export interface AgentTextChunkEvent extends BaseEvent {
  event_type: "agent_text_chunk";
  cascade_id: string;
  agent_label: string;
  text: string;
}

// ── Heartbeat events ───────────────────────────────────────────────

export interface HeartbeatStartedEvent extends BaseEvent {
  event_type: "heartbeat_started";
  heartbeat_run_id: string;
}

export interface HeartbeatTierCompletedEvent extends BaseEvent {
  event_type: "heartbeat_tier_completed";
  heartbeat_run_id: string;
  tier: number;
  outcome: string;
  escalated: boolean;
}

export interface HeartbeatCompletedEvent extends BaseEvent {
  event_type: "heartbeat_completed";
  heartbeat_run_id: string;
  highest_tier: number;
}

// ── Seed progress events ──────────────────────────────────────────

export interface SeedStartedEvent extends BaseEvent {
  event_type: "seed_started";
  layer_slug: string;
  agent_label: string;
}

export interface SeedProgressEvent extends BaseEvent {
  event_type: "seed_progress";
  layer_slug: string;
  step: string;
  message: string;
}

export interface SeedCompletedEvent extends BaseEvent {
  event_type: "seed_completed";
  layer_slug: string;
  success: boolean;
  message: string;
  items_created: number;
}

// ── Idea events ────────────────────────────────────────────────────

export interface IdeaSubmittedEvent extends BaseEvent {
  event_type: "idea_submitted";
  idea_id: string;
  source: string;
}

export interface IdeaEvaluationStartedEvent extends BaseEvent {
  event_type: "idea_evaluation_started";
  idea_id: string;
}

export interface IdeaEvaluationCompletedEvent extends BaseEvent {
  event_type: "idea_evaluation_completed";
  idea_id: string;
}

export interface IdeaGenerationStartedEvent extends BaseEvent {
  event_type: "idea_generation_started";
}

export interface IdeaGenerationCompletedEvent extends BaseEvent {
  event_type: "idea_generation_completed";
  count: number;
}

// ── System events ──────────────────────────────────────────────────

export interface UserLoginEvent extends BaseEvent {
  event_type: "user_login";
  email: string;
}

export interface UserCreatedEvent extends BaseEvent {
  event_type: "user_created";
  email: string;
  role: string;
}

export interface CascadeLockAcquiredEvent extends BaseEvent {
  event_type: "cascade_lock_acquired";
  cascade_id: string;
}

export interface CascadeLockReleasedEvent extends BaseEvent {
  event_type: "cascade_lock_released";
  cascade_id: string;
}

// ── Conversation events ────────────────────────────────────────────

export interface ConversationStartedEvent extends BaseEvent {
  event_type: "conversation_started";
  conversation_id: string;
}

export interface ConversationTextChunkEvent extends BaseEvent {
  event_type: "conversation_text_chunk";
  conversation_id: string;
  text: string;
}

export interface ConversationFileEditEvent extends BaseEvent {
  event_type: "conversation_file_edit";
  conversation_id: string;
  layer_slug: string;
  filename: string;
  action: "write" | "delete";
}

export interface ConversationTurnCompleteEvent extends BaseEvent {
  event_type: "conversation_turn_complete";
  conversation_id: string;
  message_id: string;
  files_edited: string[];
}

export interface ConversationCascadePendingEvent extends BaseEvent {
  event_type: "conversation_cascade_pending";
  conversation_id: string;
  starting_layer: string;
}

export interface ConversationTurnErrorEvent extends BaseEvent {
  event_type: "conversation_turn_error";
  conversation_id: string;
  error_message: string;
}

// ── Discriminated union ────────────────────────────────────────────

export type PolicyEvent =
  // Cascade lifecycle
  | CascadeStartedEvent
  | CascadeCompletedEvent
  | CascadeFailedEvent
  | CascadePausedEvent
  | CascadeResumedEvent
  | CascadeCancelledEvent
  | CascadeQueuedEvent
  // Layer processing
  | LayerGenerationStartedEvent
  | LayerGenerationCompletedEvent
  | CriticStartedEvent
  | CriticCompletedEvent
  | SynthesisStartedEvent
  | SynthesisCompletedEvent
  // Agent streaming
  | AgentTextChunkEvent
  // Heartbeat
  | HeartbeatStartedEvent
  | HeartbeatTierCompletedEvent
  | HeartbeatCompletedEvent
  // Seed progress
  | SeedStartedEvent
  | SeedProgressEvent
  | SeedCompletedEvent
  // Ideas
  | IdeaSubmittedEvent
  | IdeaEvaluationStartedEvent
  | IdeaEvaluationCompletedEvent
  | IdeaGenerationStartedEvent
  | IdeaGenerationCompletedEvent
  // System
  | UserLoginEvent
  | UserCreatedEvent
  | CascadeLockAcquiredEvent
  | CascadeLockReleasedEvent
  // Conversation
  | ConversationStartedEvent
  | ConversationTextChunkEvent
  | ConversationFileEditEvent
  | ConversationTurnCompleteEvent
  | ConversationCascadePendingEvent
  | ConversationTurnErrorEvent;

// ── Replay API response ────────────────────────────────────────────

export interface ReplayEvent {
  id: number; // Database integer ID
  event_type: EventType;
  timestamp: string;
  data: Record<string, unknown>;
  layer_slug: string | null;
  category: EventCategory | null;
}

export interface ReplayResponse {
  events: ReplayEvent[];
  since_id: number;
  overflow: boolean;
}

export interface ActivityResponse {
  events: ReplayEvent[];
  limit: number;
  offset: number;
}

// ── Category mapping ───────────────────────────────────────────────

const EVENT_CATEGORY_MAP: Record<EventType, EventCategory> = {
  cascade_started: "cascade",
  cascade_completed: "cascade",
  cascade_failed: "cascade",
  cascade_paused: "cascade",
  cascade_resumed: "cascade",
  cascade_cancelled: "cascade",
  cascade_queued: "cascade",
  layer_generation_started: "cascade",
  layer_generation_completed: "cascade",
  critic_started: "cascade",
  critic_completed: "cascade",
  synthesis_started: "cascade",
  synthesis_completed: "cascade",
  agent_text_chunk: "cascade",
  heartbeat_started: "heartbeat",
  heartbeat_tier_completed: "heartbeat",
  heartbeat_completed: "heartbeat",
  seed_started: "seed",
  seed_progress: "seed",
  seed_completed: "seed",
  idea_submitted: "idea",
  idea_evaluation_started: "idea",
  idea_evaluation_completed: "idea",
  idea_generation_started: "idea",
  idea_generation_completed: "idea",
  user_login: "system",
  user_created: "system",
  cascade_lock_acquired: "system",
  cascade_lock_released: "system",
  // Conversation
  conversation_started: "conversation",
  conversation_text_chunk: "conversation",
  conversation_file_edit: "conversation",
  conversation_turn_complete: "conversation",
  conversation_cascade_pending: "conversation",
  conversation_turn_error: "conversation",
};

export function getEventCategory(eventType: EventType): EventCategory {
  return EVENT_CATEGORY_MAP[eventType];
}

File: src/policy_factory/events.py
"""Event system for Policy Factory — typed events and async EventEmitter.

Provides:
- Typed dataclass events for all event categories (cascade, layer, agent,
  heartbeat, ideas, system).
- An EventEmitter with async-safe pub/sub for decoupled event dispatch.

Follows the cc-runner EventEmitter pattern with Policy Factory-specific
event types.
"""

from __future__ import annotations

import asyncio
import logging
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable, Literal

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Event type literals
# ---------------------------------------------------------------------------

EventCategory = Literal["cascade", "conversation", "heartbeat", "idea", "seed", "system"]

EventType = Literal[
    # Cascade lifecycle (7)
    "cascade_started",
    "cascade_completed",
    "cascade_failed",
    "cascade_paused",
    "cascade_resumed",
    "cascade_cancelled",
    "cascade_queued",
    # Layer processing (6)
    "layer_generation_started",
    "layer_generation_completed",
    "critic_started",
    "critic_completed",
    "synthesis_started",
    "synthesis_completed",
    # Agent streaming (1)
    "agent_text_chunk",
    # Heartbeat (3)
    "heartbeat_started",
    "heartbeat_tier_completed",
    "heartbeat_completed",
    # Seed progress (3)
    "seed_started",
    "seed_progress",
    "seed_completed",
    # Ideas (5)
    "idea_submitted",
    "idea_evaluation_started",
    "idea_evaluation_completed",
    "idea_generation_started",
    "idea_generation_completed",
    # System (4)
    "user_login",
    "user_created",
    "cascade_lock_acquired",
    "cascade_lock_released",
    # Conversation (6)
    "conversation_started",
    "conversation_text_chunk",
    "conversation_file_edit",
    "conversation_turn_complete",
    "conversation_cascade_pending",
    "conversation_turn_error",
]

# Mapping from event_type → category for efficient filtering
_EVENT_CATEGORY_MAP: dict[str, EventCategory] = {
    # Cascade lifecycle
    "cascade_started": "cascade",
    "cascade_completed": "cascade",
    "cascade_failed": "cascade",
    "cascade_paused": "cascade",
    "cascade_resumed": "cascade",
    "cascade_cancelled": "cascade",
    "cascade_queued": "cascade",
    # Layer processing (part of cascade)
    "layer_generation_started": "cascade",
    "layer_generation_completed": "cascade",
    "critic_started": "cascade",
    "critic_completed": "cascade",
    "synthesis_started": "cascade",
    "synthesis_completed": "cascade",
    # Agent streaming (part of cascade)
    "agent_text_chunk": "cascade",
    # Heartbeat
    "heartbeat_started": "heartbeat",
    "heartbeat_tier_completed": "heartbeat",
    "heartbeat_completed": "heartbeat",
    # Seed
    "seed_started": "seed",
    "seed_progress": "seed",
    "seed_completed": "seed",
    # Ideas
    "idea_submitted": "idea",
    "idea_evaluation_started": "idea",
    "idea_evaluation_completed": "idea",
    "idea_generation_started": "idea",
    "idea_generation_completed": "idea",
    # System
    "user_login": "system",
    "user_created": "system",
    "cascade_lock_acquired": "system",
    "cascade_lock_released": "system",
    # Conversation
    "conversation_started": "conversation",
    "conversation_text_chunk": "conversation",
    "conversation_file_edit": "conversation",
    "conversation_turn_complete": "conversation",
    "conversation_cascade_pending": "conversation",
    "conversation_turn_error": "conversation",
}


def get_event_category(event_type: str) -> EventCategory | None:
    """Return the category for a given event type string."""
    return _EVENT_CATEGORY_MAP.get(event_type)


# ---------------------------------------------------------------------------
# Base event
# ---------------------------------------------------------------------------


@dataclass
class BaseEvent:
    """Base class for all Policy Factory events.

    Each event has:
    - A unique ``id`` (UUID, generated on creation) for client deduplication.
    - A ``timestamp`` (UTC datetime, serialised to ISO 8601).
    - An ``event_type`` string set by each subclass (not passed by caller).
    """

    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    event_type: str = field(init=False)  # Set by subclass

    def to_dict(self) -> dict[str, Any]:
        """Return a JSON-serialisable dictionary of this event."""
        return {
            "id": self.id,
            "event_type": self.event_type,
            "timestamp": self.timestamp.isoformat(),
        }


# ---------------------------------------------------------------------------
# Cascade lifecycle events (7)
# ---------------------------------------------------------------------------


@dataclass
class CascadeStarted(BaseEvent):
    """Emitted when a cascade begins."""

    cascade_id: str = ""
    trigger_source: str = ""  # user_input, layer_refresh, heartbeat, seed
    starting_layer: str = ""
    event_type: str = field(default="cascade_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "trigger_source": self.trigger_source,
            "starting_layer": self.starting_layer,
        }


@dataclass
class CascadeCompleted(BaseEvent):
    """Emitted when a cascade completes successfully."""

    cascade_id: str = ""
    success: bool = True
    event_type: str = field(default="cascade_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "success": self.success,
        }


@dataclass
class CascadeFailed(BaseEvent):
    """Emitted when a cascade fails."""

    cascade_id: str = ""
    error: str = ""
    failed_layer: str = ""
    failed_step: str = ""  # generation, critics, synthesis
    event_type: str = field(default="cascade_failed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "error": self.error,
            "failed_layer": self.failed_layer,
            "failed_step": self.failed_step,
        }


@dataclass
class CascadePaused(BaseEvent):
    """Emitted when a cascade is paused due to error."""

    cascade_id: str = ""
    error: str = ""
    paused_layer: str = ""
    paused_step: str = ""
    event_type: str = field(default="cascade_paused", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "error": self.error,
            "paused_layer": self.paused_layer,
            "paused_step": self.paused_step,
        }


@dataclass
class CascadeResumed(BaseEvent):
    """Emitted when a paused cascade is resumed."""

    cascade_id: str = ""
    event_type: str = field(default="cascade_resumed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
        }


@dataclass
class CascadeCancelled(BaseEvent):
    """Emitted when a cascade is cancelled."""

    cascade_id: str = ""
    event_type: str = field(default="cascade_cancelled", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
        }


@dataclass
class CascadeQueued(BaseEvent):
    """Emitted when a cascade is added to the queue."""

    cascade_id: str = ""
    queue_position: int = 0
    event_type: str = field(default="cascade_queued", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "queue_position": self.queue_position,
        }


# ---------------------------------------------------------------------------
# Layer processing events (6)
# ---------------------------------------------------------------------------


@dataclass
class LayerGenerationStarted(BaseEvent):
    """Emitted when layer generation begins."""

    cascade_id: str = ""
    layer_slug: str = ""
    event_type: str = field(default="layer_generation_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "layer_slug": self.layer_slug,
        }


@dataclass
class LayerGenerationCompleted(BaseEvent):
    """Emitted when layer generation completes."""

    cascade_id: str = ""
    layer_slug: str = ""
    event_type: str = field(default="layer_generation_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "layer_slug": self.layer_slug,
        }


@dataclass
class CriticStarted(BaseEvent):
    """Emitted when a critic agent starts."""

    cascade_id: str = ""
    layer_slug: str = ""
    critic_archetype: str = ""  # e.g. "realist", "liberal-institutionalist"
    event_type: str = field(default="critic_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "layer_slug": self.layer_slug,
            "critic_archetype": self.critic_archetype,
        }


@dataclass
class CriticCompleted(BaseEvent):
    """Emitted when a critic agent completes."""

    cascade_id: str = ""
    layer_slug: str = ""
    critic_archetype: str = ""
    event_type: str = field(default="critic_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "layer_slug": self.layer_slug,
            "critic_archetype": self.critic_archetype,
        }


@dataclass
class SynthesisStarted(BaseEvent):
    """Emitted when synthesis begins."""

    cascade_id: str = ""
    layer_slug: str = ""
    event_type: str = field(default="synthesis_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "layer_slug": self.layer_slug,
        }


@dataclass
class SynthesisCompleted(BaseEvent):
    """Emitted when synthesis completes."""

    cascade_id: str = ""
    layer_slug: str = ""
    event_type: str = field(default="synthesis_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "layer_slug": self.layer_slug,
        }


# ---------------------------------------------------------------------------
# Agent streaming events (1)
# ---------------------------------------------------------------------------


@dataclass
class AgentTextChunk(BaseEvent):
    """High-frequency event carrying streamed agent reasoning text."""

    cascade_id: str = ""  # Or other context ID
    agent_label: str = ""  # Human-readable, e.g. "Realist critic"
    text: str = ""
    event_type: str = field(default="agent_text_chunk", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
            "agent_label": self.agent_label,
            "text": self.text,
        }


# ---------------------------------------------------------------------------
# Heartbeat events (3)
# ---------------------------------------------------------------------------


@dataclass
class HeartbeatStarted(BaseEvent):
    """Emitted when a heartbeat run begins."""

    heartbeat_run_id: str = ""
    event_type: str = field(default="heartbeat_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "heartbeat_run_id": self.heartbeat_run_id,
        }


@dataclass
class HeartbeatTierCompleted(BaseEvent):
    """Emitted when a heartbeat tier completes."""

    heartbeat_run_id: str = ""
    tier: int = 0
    outcome: str = ""
    escalated: bool = False
    event_type: str = field(default="heartbeat_tier_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "heartbeat_run_id": self.heartbeat_run_id,
            "tier": self.tier,
            "outcome": self.outcome,
            "escalated": self.escalated,
        }


@dataclass
class HeartbeatCompleted(BaseEvent):
    """Emitted when a heartbeat run completes."""

    heartbeat_run_id: str = ""
    highest_tier: int = 0
    event_type: str = field(default="heartbeat_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "heartbeat_run_id": self.heartbeat_run_id,
            "highest_tier": self.highest_tier,
        }


# ---------------------------------------------------------------------------
# Seed progress events (3)
# ---------------------------------------------------------------------------


@dataclass
class SeedStarted(BaseEvent):
    """Emitted when a seed operation begins for a layer."""

    layer_slug: str = ""
    agent_label: str = ""
    event_type: str = field(default="seed_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "layer_slug": self.layer_slug,
            "agent_label": self.agent_label,
        }


@dataclass
class SeedProgress(BaseEvent):
    """Emitted at key milestones during a seed operation."""

    layer_slug: str = ""
    step: str = ""  # agent_running, parsing, writing, committing, cascade
    message: str = ""
    event_type: str = field(default="seed_progress", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "layer_slug": self.layer_slug,
            "step": self.step,
            "message": self.message,
        }


@dataclass
class SeedCompleted(BaseEvent):
    """Emitted when a seed operation finishes (success or failure)."""

    layer_slug: str = ""
    success: bool = True
    message: str = ""
    items_created: int = 0
    event_type: str = field(default="seed_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "layer_slug": self.layer_slug,
            "success": self.success,
            "message": self.message,
            "items_created": self.items_created,
        }


# ---------------------------------------------------------------------------
# Idea events (5)
# ---------------------------------------------------------------------------


@dataclass
class IdeaSubmitted(BaseEvent):
    """Emitted when an idea is submitted."""

    idea_id: str = ""
    source: str = ""  # "human" or "AI"
    event_type: str = field(default="idea_submitted", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "idea_id": self.idea_id,
            "source": self.source,
        }


@dataclass
class IdeaEvaluationStarted(BaseEvent):
    """Emitted when idea evaluation begins."""

    idea_id: str = ""
    event_type: str = field(default="idea_evaluation_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "idea_id": self.idea_id,
        }


@dataclass
class IdeaEvaluationCompleted(BaseEvent):
    """Emitted when idea evaluation completes."""

    idea_id: str = ""
    event_type: str = field(default="idea_evaluation_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "idea_id": self.idea_id,
        }


@dataclass
class IdeaGenerationStarted(BaseEvent):
    """Emitted when batch idea generation begins."""

    event_type: str = field(default="idea_generation_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {**super().to_dict()}


@dataclass
class IdeaGenerationCompleted(BaseEvent):
    """Emitted when batch idea generation completes."""

    count: int = 0
    event_type: str = field(default="idea_generation_completed", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "count": self.count,
        }


# ---------------------------------------------------------------------------
# System events (4)
# ---------------------------------------------------------------------------


@dataclass
class UserLogin(BaseEvent):
    """Emitted when a user logs in."""

    email: str = ""
    event_type: str = field(default="user_login", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "email": self.email,
        }


@dataclass
class UserCreated(BaseEvent):
    """Emitted when a user account is created."""

    email: str = ""
    role: str = ""
    event_type: str = field(default="user_created", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "email": self.email,
            "role": self.role,
        }


@dataclass
class CascadeLockAcquired(BaseEvent):
    """Emitted when the cascade lock is acquired."""

    cascade_id: str = ""
    event_type: str = field(default="cascade_lock_acquired", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
        }


@dataclass
class CascadeLockReleased(BaseEvent):
    """Emitted when the cascade lock is released."""

    cascade_id: str = ""
    event_type: str = field(default="cascade_lock_released", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "cascade_id": self.cascade_id,
        }


# ---------------------------------------------------------------------------
# Conversation events (6)
# ---------------------------------------------------------------------------


@dataclass
class ConversationStarted(BaseEvent):
    """Emitted when a conversation turn begins processing."""

    conversation_id: str = ""
    event_type: str = field(default="conversation_started", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "conversation_id": self.conversation_id,
        }


@dataclass
class ConversationTextChunk(BaseEvent):
    """Emitted during streaming response generation."""

    conversation_id: str = ""
    text: str = ""
    event_type: str = field(default="conversation_text_chunk", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "conversation_id": self.conversation_id,
            "text": self.text,
        }


@dataclass
class ConversationFileEdit(BaseEvent):
    """Emitted when the agent writes or deletes a file."""

    conversation_id: str = ""
    layer_slug: str = ""
    filename: str = ""
    action: str = ""  # "write" or "delete"
    event_type: str = field(default="conversation_file_edit", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "conversation_id": self.conversation_id,
            "layer_slug": self.layer_slug,
            "filename": self.filename,
            "action": self.action,
        }


@dataclass
class ConversationTurnComplete(BaseEvent):
    """Emitted when the agent finishes its turn."""

    conversation_id: str = ""
    message_id: str = ""
    files_edited: list[str] = field(default_factory=list)  # layer_slug/filename paths
    event_type: str = field(default="conversation_turn_complete", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "conversation_id": self.conversation_id,
            "message_id": self.message_id,
            "files_edited": self.files_edited,
        }


@dataclass
class ConversationCascadePending(BaseEvent):
    """Emitted when conversation edits to foundational layers queue a cascade."""

    conversation_id: str = ""
    starting_layer: str = ""
    event_type: str = field(default="conversation_cascade_pending", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "conversation_id": self.conversation_id,
            "starting_layer": self.starting_layer,
        }


@dataclass
class ConversationTurnError(BaseEvent):
    """Emitted when the agent turn fails."""

    conversation_id: str = ""
    error_message: str = ""
    event_type: str = field(default="conversation_turn_error", init=False)

    def to_dict(self) -> dict[str, Any]:
        return {
            **super().to_dict(),
            "conversation_id": self.conversation_id,
            "error_message": self.error_message,
        }


# ---------------------------------------------------------------------------
# EventEmitter
# ---------------------------------------------------------------------------

# Type alias for event handlers (sync or async)
EventHandler = Callable[[BaseEvent], None] | Callable[[BaseEvent], Awaitable[None]]


class EventEmitter:
    """Async-safe pub/sub for event dispatch.

    Follows the cc-runner pattern: handlers can be sync or async callables.
    A failing handler is logged but does not prevent other handlers from
    receiving the event.
    """

    def __init__(self) -> None:
        self._handlers: list[EventHandler] = []

    def subscribe(self, handler: EventHandler) -> None:
        """Subscribe a handler to receive events."""
        self._handlers.append(handler)

    def unsubscribe(self, handler: EventHandler) -> None:
        """Unsubscribe a handler from receiving events."""
        try:
            self._handlers.remove(handler)
        except ValueError:
            pass  # Handler not found — no-op

    async def emit(self, event: BaseEvent) -> None:
        """Emit an event to all subscribed handlers.

        Awaits async handlers. If a handler raises an exception, it is
        logged but does not propagate — other handlers still receive
        the event.
        """
        for handler in list(self._handlers):
            try:
                result = handler(event)
                if asyncio.iscoroutine(result):
                    await result
            except Exception:
                logger.exception(
                    "Event handler %r failed for event %s",
                    handler,
                    event.event_type,
                )
